### PR TITLE
Docker compose implementation

### PR DESCRIPTION
**Pull request by warezgibzzz**
TL;DR: Use docker compose to get an working environment.

I will write status updates to #85.

### CODE REVIEW

File: docker-compose.full.yml [hunk 1]
```diff
@@ -0,0 +1,82 @@
+version: '2'
```
Comment (leonprou) [inline docker-compose.full.yml]: hey, can you use version 3? I've read it's the recommendation https://docs.docker.com/compose/compose-file/compose-versioning/

Comment (warezgibzzz) [reply docker-compose.full.yml]: sure, will do.

File: server/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,6 @@
+FROM node:latest
+RUN apt install git
+COPY package.json .
+COPY package-lock.json .
+RUN npm install
+COPY . .
```
Comment (leonprou) [inline server/Dockerfile]: When I listed the files in the containers the app's directories was in the root directory. Can you move them to some arbitrary path, like `/code`.

File: docker-compose.full.yml [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                    - $PROJECT_NAME-mongo
+        ports:
+            - "27017"
+    ipfs:
+        container_name: $PROJECT_NAME-ipfs
+        image: ipfs/go-ipfs:latest
+        user: "${UID}:${GID}"
+        restart: always
+        env_file: .env
+        networks:
```
Comment (leonprou) [inline docker-compose.full.yml]: add this ports (used by ipfs to sync with the swarm)

        ports:
          - "4001"
          - "5001"

File: .env.dist [hunk 4]
```diff
@@ -0,0 +1,32 @@
+# Project name - affects containers and network name
+PROJECT_NAME=cln-community
+
+# User/group ID management
+UID=1000
+GID=1000
+
+# Application parameters
+SERVER_SECRET=somerandomhash
```
[the remark this answers is not in the document]
Comment (warezgibzzz) [reply .env.dist]: except UID and GID - it's docker specific

File: docker-compose.full.yml [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    mongo:
+        container_name: $PROJECT_NAME-mongo
+        build: ./docker/mongo
+        restart: always
+        env_file: .env
+        environment:
+            - MONGO_INITDB_ROOT_USERNAME=${MONGO_USER}
+            - MONGO_INITDB_ROOT_PASSWORD=${MONGO_PASS}
+            - MONGO_INITDB_DATABASE=${MONGO_DBNAME}
+        environment:
```
[the remark this answers is not in the document]
Comment (warezgibzzz) [reply docker-compose.full.yml]: fixed

File: docker-compose.full.yml (hunk 3 above)
Comment (warezgibzzz) [reply docker-compose.full.yml]: fixed

File: docker-compose.full.yml [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        env_file: .env
+        command: geth --datadir=/geth --testnet --rpc --rpcaddr 0.0.0.0 --rpcapi="db,eth,net,web3,personal,web3"
+        networks:
+            internal-network:
+            external-network:
+                aliases:
+                    - $PROJECT_NAME-geth
+        ports:
+            - "8545"
+        depends_on:
```
[the remark this answers is not in the document]
Comment (warezgibzzz) [reply docker-compose.full.yml]: geth ommited now, so, fixed :)

File: docker-compose.full.yml [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+            - ipfs
+        volumes:
+            - ./docker/geth/:/geth
+    server:
+        container_name: $PROJECT_NAME-server
+        build: ./server
+        user: "${UID}:${GID}"
+        restart: always
+        env_file: .env
+        depends_on:
```
[the remark this answers is not in the document]
Comment (warezgibzzz) [reply docker-compose.full.yml]: done

File: docker-compose.full.yml [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+        ports:
+            - "8081:$APP_PORT"
+        environment:
+            - NODE_ENV=${APP_ENV}
+            - PORT=${APP_PORT}
+            - COMMUNITY_MONGO_URI=mongodb://mongo/${MONGO_DBNAME}
+            - COMMUNITY_MONGO_USER=${MONGO_USER}
+            - COMMUNITY_MONGO_PASS=${MONGO_PASS}
+            - COMMUNITY_MONGO_AUTH_SOURCE=${MONGO_AUTH_SOURCE}
+            - COMMUNITY_IPFS_HOST=ipfs
```
[the remark this answers is not in the document]
Comment (warezgibzzz) [reply docker-compose.full.yml]: done

File: docker-compose.full.yml [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+            - COMMUNITY_IPFS_PORT=${IPFS_PORT}
+            - COMMUNITY_IPFS_PROTOCOL=${IPFS_PROTO}
+            - COMMUNITY_IPFS_TIMEOUT=${IPFS_READ_TIMEOUT}
+            - COMMUNITY_API_SECRET=${SERVER_SECRET}
+            - COMMUNITY_MANDRILL_API_KEY=${MANDRILL_API}
+            - COMMUNITY_MANDRILL_SEND_TO=${MANDRILL_SEND_TO}
+            - COMMUNITY_MAILCHIMP_API_BASE=${MAILCHIMP_API_BASE}
+            - COMMUNITY_MAILCHIMP_LIST=${MAILCHIMP_LIST_NAME}
+            - COMMUNITY_AMAZON_BUCKET_API=${AMAZON_BUCKET_API}
+            - COMMUNITY_WEB3_PROVIDER=http://geth:8545
```
[the remark this answers is not in the document]
Comment (warezgibzzz) [reply docker-compose.full.yml]: done

File: server/Dockerfile (hunk 2 above)
Comment (warezgibzzz) [reply server/Dockerfile]: moved to /app

File: .gitignore [hunk 10]
```diff
@@ -63,3 +63,16 @@ typings/
...
+# Docker local data
+docker-compose.yml
+.env
+
+# Docker volumes data - depends on user, dont want these in repo.
+docker/mongo/config/
+docker/mongo/db/
+docker/ipfs/data/
+docker/ipfs/export/
+docker/geth/
```
Comment (leonprou) [inline .gitignore]: geth not needed

File: client/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,8 @@
+FROM node:latest
+USER $UID:$GID
+WORKDIR /app
+RUN apt install git
```
Comment (leonprou) [inline client/Dockerfile]: do git needed?

File: docker-compose.full.yml [hunk 12]
```diff
@@ -0,0 +1,108 @@
+version: '3.2'
```
Comment (leonprou) [inline docker-compose.full.yml]: this file can renamed to docker-compose.yml, right?

File: server/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,8 @@
+FROM node:latest
+USER $UID:$GID
+WORKDIR /app
+RUN apt install git
```
Comment (leonprou) [inline server/Dockerfile]: is git needed?

File: client/Dockerfile (hunk 11 above)
Comment (warezgibzzz) [reply client/Dockerfile]: Yep. Some npm packages can fetch through git

File: server/Dockerfile (hunk 13 above)
Comment (warezgibzzz) [reply server/Dockerfile]: Yep. Some npm packages can fetch through git

File: docker-compose.full.yml (hunk 12 above)
Comment (warezgibzzz) [reply docker-compose.full.yml]: So, yea. that file can be copied or linked by ln -s to docker-compose.yml

**Pull request merged**